pkg/clustercert/cert/generator.go

```diff
@@ -0,0 +1,296 @@
+// Copyright © 2022 Alibaba Group Holding Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cert
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math"
+	"math/big"
+	"net"
+	"os"
+	"path"
+	"path/filepath"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/util/keyutil"
+)
+
+const (
+	// PrivateKeyBlockType is a possible value for pem.Block.Type.
+	PrivateKeyBlockType = "PRIVATE KEY"
+	// PublicKeyBlockType is a possible value for pem.Block.Type.
+	PublicKeyBlockType = "PUBLIC KEY"
+	// CertificateBlockType is a possible value for pem.Block.Type.
+	CertificateBlockType = "CERTIFICATE"
+	// RSAPrivateKeyBlockType is a possible value for pem.Block.Type.
+	RSAPrivateKeyBlockType = "RSA PRIVATE KEY"
+	rsaKeySize             = 2048
+	duration365d           = time.Hour * 24 * 365
+)
+
+// KeyPairFileGenerator write symmetric encryption key, like: sa.key and sa.pub
+type KeyPairFileGenerator struct {
+	path string
+	name string
+}
+
+func NewKeyPairFileGenerator(certPath string, certName string) KeyPairFileGenerator {
+	return KeyPairFileGenerator{
+		path: certPath,
+		name: certName,
+	}
+}
+
+func (k KeyPairFileGenerator) GenerateAll() error {
+	_, err := os.Stat(path.Join(k.path, "sa.key"))
+	if !os.IsNotExist(err) {
+		logrus.Info("sa.key sa.pub already exist")
+		return nil
+	}
+
+	key, err := NewPrivateKey(x509.RSA)
+	if err != nil {
+		return err
+	}
+
+	err = k.writePrivateKey(key)
+	if err != nil {
+		return err
+	}
+
+	return k.writePublicKey(key.Public())
+}
+
+func (k KeyPairFileGenerator) writePrivateKey(key crypto.Signer) error {
+	if key == nil {
+		return errors.New("private key cannot be nil when writing to file")
+	}
+
+	privateKeyPath := PathForKey(k.path, k.name)
+	encoded, err := keyutil.MarshalPrivateKeyToPEM(key)
+	if err != nil {
+		return fmt.Errorf("unable to marshal private key to PEM %v", err)
+	}
+	if err := keyutil.WriteKey(privateKeyPath, encoded); err != nil {
+		return fmt.Errorf("unable to write private key to file %s %v", privateKeyPath, err)
```

Comment:
Please use `:` to separate the error wrap information and the detailed error. This would bring much more readability for sealer end users.